### ISSUE TITLE
Container.Verify() does not dispose WebRequestLifestyle instances.

### ISSUE THREAD

**Issue by ryletko** [simpleinjector/SimpleInjector#95]
Hello! When I use RegisterPerWebRequest and then do Verify(), it creates an instance of type and doesn't dispose it. The instance is alive all time while the application is working. I use WebApi and WebForms injections together, other instances of that type (after Verify) are created and disposed correctly according to 'perwebrequest' lifestyle, but the first instance that was created in Verify method is not disposed at all.

**Comment by dotnetjunkie** [comment]
Which version of Simple Injector are you using?

**Comment by ryletko** [comment]
tested on 2.7.3 and 2.8.3

**Comment by dotnetjunkie** [comment]
I've been able to reproduce this. What is happening is the following: during verification, Simple Injector will create a special scope that is active during verification. `ScopedLifestyle` derivatives (such as the `WebRequestLifestyle`) will use that scope in case their own scope isn't active. During web application startup however, the `WebRequestLifestyle` will return a scope, because `HttpContext.Current` returns a value. If it was `null`, the verification scope would have been used.

Because the `WebRequestLifestyle` will return a real scope, the scoped instances are cached in the `HttpContext.Items` dictionary. When the verification process ends, `Verify()` will dispose the verification scope, but this scope is empty.

Typically, the disposable items would now be disposed when the web request ends, but unfortunately, the `HttpContext.Current` instance that is returned during application startup is a fake request. The `EndRequest` event will therefore never be called and this causes those instances to not be disposed.

This is a design flaw. We'll investigate in how to solve this.

Normally, it shouldn't be an issue that such single instance isn't disposed. Classes that contain references to managed resources should implement a finalizer that will eventually clean up the resource. 

If however, it is a real problem that this class isn't disposed, what you can do as a work around is resolving this instance directly after calling `Verify()` and dispose it manually as follows:

``` c#
container.Verifty();
container.GetInstance<MyDisposableClass>().Dispose();
```

Or you can ensure that `Verify()` is only called in debug mode:


``` c#
#if DEBUG
container.GetInstance<MyDisposableClass>().Dispose();
#endif
```

Or you can move the call to `Verify()` out of the startup path, into an integration test. During an integration test it shouldn't be a problem at all.

Or in case you have many disposable instances that need disposal, you can call the internal `WebRequestLifestyle.CleanUpWebRequest` during reflection after the call to `Verify()`.

**Comment by abatishchev** [comment]
I personally moved Verify() to an integration test what decreased a hassle around the difference between real-world run-time container behavior and during testing/verification/etc.

**Comment by ryletko** [comment]
Thank you for your reply. That's a good point about moving Verify() to integration tests.

**Comment by dotnetjunkie** [comment]
The bug has been fixed in [v3.0.3](https://github.com/simpleinjector/SimpleInjector/releases/tag/v3.0.3). This new patch release has been pushed to NuGet.

Thanks for taking the time to report this!

**Comment by trnktms** [comment]
I have the same problem with version 3.1.0. Is it possible that this fix removed from 3.1.0 release?

**Comment by trnktms** [comment]
Thank you for your quick response @dotnetjunkie!
As a quick fix I just used the solution what you suggested above:
```
#if DEBUG
container.Verify();
#endif
```
